### PR TITLE
Use bashunit for e2e tests

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

File: .github/workflows/e2e-tests.yml
Comment (staabm) [inline .github/workflows/e2e-tests.yml]: do we still want a single isolated github action job per test?

File: e2e/result-cache-1/test.sh
Comment (staabm) [inline e2e/result-cache-1/test.sh]: I need to redirect stderr to stdout for all commands of the test because bashunit reports a test-failure as soon as something is written to ~~stdout~~ stderr

**Comment by ondrejmirtes** [comment]
I'd rather install it with a script for a specific version than had it commited https://bashunit.typeddevs.com/installation

File: e2e/result-cache-1/test.sh
Comment (ondrejmirtes) [reply e2e/result-cache-1/test.sh]: I think I want to ask the authors what's the best way to do it here.

File: .github/workflows/e2e-tests.yml
Comment (ondrejmirtes) [reply .github/workflows/e2e-tests.yml]: Yes, because I'm worried about polluting the state and making the tests worthless.

Comment (ondrejmirtes) [reply .github/workflows/e2e-tests.yml]: It'd be great if we could just use the asserts outside of running with bashunit.

Comment (staabm) [reply .github/workflows/e2e-tests.yml]: > It'd be great if we could just use the asserts outside of running with bashunit.

do you mean we should leave this script section in the e2e-test.yml file as before but just replacing `grep` with `assert_contains`? should I not add `test.js` like files in the test folders?

Comment (ondrejmirtes) [reply .github/workflows/e2e-tests.yml]: Yes that's exactly what I wanted to do, but I don't know if bashunit supports that.

Comment (staabm) [reply .github/workflows/e2e-tests.yml]: so something like

```diff
              cd e2e/trait-caching
              ../../bin/phpstan analyze --no-progress --level 8 --error-format raw data/
              patch -b data/TraitOne.php < TraitOne.patch
              OUTPUT=$(../../bin/phpstan analyze --no-progress --level 8 --error-format raw data/ || true)
              echo "$OUTPUT"
              [ $(echo "$OUTPUT" | wc -l) -eq 1 ]
-             grep 'Method TraitsCachingIssue\\TestClassUsingTrait::doBar() should return stdClass but returns Exception.' <<< "$OUTPUT"
+             bashunit assert_contains 'Method TraitsCachingIssue\\TestClassUsingTrait::doBar() should return stdClass but returns Exception.' "$OUTPUT"

```

..?

Comment (staabm) [reply .github/workflows/e2e-tests.yml]: lets see https://github.com/TypedDevs/bashunit/issues/257

File: e2e/assert.sh
Comment (staabm) [inline e2e/assert.sh]: this script loads `bashunit` and allows to call [the `assert*` bash-functions of `bashunit`](https://github.com/TypedDevs/bashunit/blob/bf86c013eb7294294237da29317ea907eb8811e5/src/assert.sh).

I made it return 0 on successfull assertion or print the `bashunit` error on failure

grafik

Comment (staabm) [reply e2e/assert.sh]: if you like the approach I could adjust remaining shell assertions in the other test cases

**Comment by ondrejmirtes** [comment]
I have no idea what the assert.sh script does. I'm not a bash expert. I'd like to hear from Bashunit creators on this.

File: e2e/assert.sh
Comment (Chemaclass) [reply e2e/assert.sh]: This is a pretty cool approach. I was thinking about doing something similar. Maybe, actually, we could have this code somehow inside bashunit itself and allow people using this feature if they would want to as well (inspired by your idea here https://github.com/phpstan/phpstan-src/pull/3160#discussion_r1640951032).

I will experiment a couple of hours in the train today, and I will give you an update later today 🏗️ 

**UPDATE**: Merged https://github.com/TypedDevs/bashunit/pull/258

File: .github/workflows/e2e-tests.yml
Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml]: Would this be what you are looking for? https://github.com/TypedDevs/bashunit/pull/258

**Comment by Chemaclass** [comment]
@ondrejmirtes @staabm you can use **Allow calling assertions outside tests** #258 by installing the beta for now (which uses main branch) -> https://bashunit.typeddevs.com/installation#install-sh and later on use the next official release (which would be out in a week or two). This way, you can start working on this and you aren't block from bashunit anymore :)

File: .github/workflows/e2e-tests.yml
Comment (staabm) [inline .github/workflows/e2e-tests.yml:214]: @Chemaclass thank you.

I have some weird case here.

```
➜  trait-caching git:(bashunit) ✗ echo "$OUTPUT"
/Users/staabm/workspace/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:15:Method TraitsCachingIssue\TestClassUsingTrait::doBar() should return stdClass but returns Exception.
/Users/staabm/workspace/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:29:Method class@anonymous/TestClassUsingTrait.php:20::doBar() should return stdClass but returns Exception.
➜  trait-caching git:(bashunit) ✗ echo "$OUTPUT" | wc -l | tr -d '[:blank:]'
2
➜  trait-caching git:(bashunit) ✗ ../bashunit -a line_count 2 "$OUTPUT"
✗ Failed: /Users/staabm/workspace/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:29:Method class@anonymous/TestClassUsingTrait.php:20::doBar() should return stdClass but returns Exception.
    Expected '/Users/staabm/workspace/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:15:Method TraitsCachingIssue\TestClassUsingTrait::doBar() should return stdClass but returns Exception.'
    to contain number of lines equal to '2'
```

do you have an idea what is wrong here?

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: I would like to reproduce it locally, so I cloned your fork (`git@github.com:Chemaclass/staabm-phpstan-src.git`), checkout your `bashunit` branch, and I get this error when running composer install - which I think it shouldn't happen as it is a virgin start (I also set php 8.1), so I guess there is a problem on the deps? Did you face the same issue, @staabm ? Any idea how to find a work around to this?

Screenshot 2024-06-17 at 21 15 14

Comment (Chemaclass) [inline .github/workflows/e2e-tests.yml:214]: I would like to reproduce it locally, so I cloned your fork (`git@github.com:Chemaclass/staabm-phpstan-src.git`), checkout your bashunit branch, and I get this error when running composer install - which I think it shouldn't happen as it is a virgin start (I also set php 8.1), so I guess there is a problem on the deps? Did you face the same issue, @staabm ? Any idea how to find a work around to this?  340425921-5b9fabbb-1df8-4442-bddc-29454deabaae

Comment (ondrejmirtes) [reply .github/workflows/e2e-tests.yml:214]: See the README.

> If you encounter dependency problem, try using `export COMPOSER_ROOT_VERSION=1.11.x-dev`

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: I think having the actual number would help to understand the issue, 
eg `to contain number of lines equal to '2' but found '1'`
because right now we know it's failing and it's not 2 but what is it? 🤔 I will try to add this now
One guess is that maybe it's only considering one line, but I am not sure, I need to debug and I would like to reproduce this bug in my local. Therefore I wrote the other comment below

**Comment by staabm** [comment]
> by installing the beta for now (which uses main branch)

I tried using the beta on my local m1 pro mac, but it didn't work.
the installer reported the "beta" was installed, but the `--help` screen does not include the `-a` option?

```
➜  curl -s https://bashunit.typeddevs.com/install.sh | bash -s ./ beta  
> Downloading non-stable version: 'beta'
> bashunit has been installed in the './' folder

➜  bashunit --help
bashunit [arguments] [options]

Arguments:
  Specifies the directory or file containing the tests to run.
  If a directory is specified, it will execute the tests within files ending with test.sh.
  If you use wildcards, bashunit will run any tests it finds.

Options:
  -f|--filter <filter>
    Filters the tests to run based on the test name.

  -s|simple || -v|verbose
    Enables simplified or verbose output to the console.

  -S|--stop-on-failure
    Force to stop the runner right after encountering one failing test.

  -e|--env <file-path>
    Load a custom env file overriding the .env environment variables.

  --version
    Displays the current version of bashunit.

  --upgrade
    Upgrade to latest version of bashunit.

  --help
    This message.

See more: https://bashunit.typeddevs.com/command-line

➜  bashunit --version
bashunit - 0.12.0

➜  bashunit -a equals "a" "b"
bashunit - 0.12.0

Tests:      0 total
Assertions: 0 total

 No tests found 
➜  bashunit -a equals "a" "a"
bashunit - 0.12.0

Tests:      0 total
Assertions: 0 total

```

File: .github/workflows/e2e-tests.yml
Comment (staabm) [reply .github/workflows/e2e-tests.yml:214]: I think I can reproduce on the CLI outside of phpstan repo with:

```
OUTPUT="one\ntwo\nthree"
➜  echo $OUTPUT
one
two
three

➜ ./bashunit -a line_count 3 "$OUTPUT"

✗ Failed: Main::exec assert
    Expected 'one\ntwo\nthree'
    to contain number of lines equal to '3'

```

Comment (staabm) [reply .github/workflows/e2e-tests.yml:214]: agree, having the actual value in the message would be helpful.

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: @staabm this could help to see the actual value https://github.com/TypedDevs/bashunit/pull/263 and enabling adding extra information to the failed assertions in case it might be needed (optional) - feel free to review/comment/approve it if you want (always! 😄) This can help us debugging the current issue after merging it and you refreshing your beta

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: Ok, in such a case the bug seems to be in the `assert_line_count` where it might consider that as 1 line and not as 3. I think that's the direction of the 🐛

Comment (staabm) [reply .github/workflows/e2e-tests.yml:214]: hmm .. the unit test regarding it looked good:
https://github.com/TypedDevs/bashunit/pull/259/files#diff-8c5ca8e706f9d9fb330cda7953d707bb4180b68f55fdff87ec30ad3f8e179ad6R339-R341

I don't know bash good enough, but I had a feeling the problem is related to the assertion via args passing

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: The issue is clearly here `actual_line_count=$(echo "$actual" | wc -l | tr -d '[:blank:]')` - still, I don't understand the issue - I mean, I don't know how to solve it yet 🤔 
Here you can see the result of the failing test:
```bash
function test_successful_assert_line_count_multiline_string_in_one_line() {
  assert_empty "$(assert_line_count 4 "one\ntwo\nthree\nfour")"
}
```

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: I got a potential fix for this ... creating PR

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: I got a potential fix for this: https://github.com/TypedDevs/bashunit/pull/264 (merged)
Can you try again to reproduce it with the latest beta, @staabm ?

Comment (staabm) [reply .github/workflows/e2e-tests.yml:214]: thanks. really appreciate your support

my initial reported case works as expected now. the build here still fails though.

I reduced it to:

```
➜  OUTPUT="/home/runner/work/phpstan-src/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:15:Method TraitsCachingIssue\TestClassUsingTrait::doBar() should return stdClass but returns Exception.
/home/runner/work/phpstan-src/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:29:Method class@anonymous/TestClassUsingTrait.php:20::doBar() should return stdClass but returns Exception."
➜  echo $OUTPUT
/home/runner/work/phpstan-src/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:15:Method TraitsCachingIssue\TestClassUsingTrait::doBar() should return stdClass but returns Exception.
/home/runner/work/phpstan-src/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:29:Method class@anonymous/TestClassUsingTrait.php:20::doBar() should return stdClass but returns Exception.
➜  ./bashunit -a line_count 2 "$OUTPUT"
✗ Failed: /home/runner/work/phpstan-src/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:29:Method class@anonymous/TestClassUsingTrait.php:20::doBar() should return stdClass but returns Exception.
    Expected '/home/runner/work/phpstan-src/phpstan-src/e2e/trait-caching/data/TestClassUsingTrait.php:15:Method TraitsCachingIssue\TestClassUsingTrait::doBar() should return stdClass but returns Exception.'
    to contain number of lines equal to '2'
    but found '1'
```

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: I can reproduce it, yes. This is a bashunit issue, I will work on it today 🏗️ 
UPDATE: It took me a couple of hours... but finally I got something out: https://github.com/TypedDevs/bashunit/pull/265

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: Alright, https://github.com/TypedDevs/bashunit/pull/265 is merged and we are getting closer to the light!
Can you try again to reproduce it with the latest beta, @staabm ?

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: With the latest [error](https://github.com/phpstan/phpstan-src/actions/runs/9567238902/job/26374488151), I think we might have found a major problem wrt multiline strings mmmm I have an idea, but it might take a while... let me see if I can create a draft for this today 🏗️ 
> UPDATE: This might be the fix for the multiline string bug: https://github.com/TypedDevs/bashunit/pull/266

Comment (Chemaclass) [reply .github/workflows/e2e-tests.yml:214]: Same drill: https://github.com/TypedDevs/bashunit/pull/266 is merged, can you try the latest beta again, @staabm? 🤞🏼

Comment (staabm) [reply .github/workflows/e2e-tests.yml:214]: @Chemaclass all e2e are green now. thank you very much for your efforts <3.

**Comment by staabm** [comment]
@ondrejmirtes this one is good to go now :)

**Review by Chemaclass — APPROVED** [review]
Awesome 🙌  For now, you can use the beta which contains the new working functionality that you need, and in the coming days, once we launch the new release, then you can pin it to the latest stable instead of beta. 
Anything you need, here we are!

**Comment by ondrejmirtes** [comment]
Awesome!

@staabm Please send an update once a stable release comes out.

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**
